File: 02/python/day2-2.py
#!/usr/bin/python

from enum import IntEnum


class Action(IntEnum):
    Rock = 1
    Paper = 2
    Scissors = 3


victories = {
    Action.Rock: [Action.Scissors],  # Rock beats scissors
    Action.Paper: [Action.Rock],  # Paper beats rock
    Action.Scissors: [Action.Paper],  # Scissors beats paper
}

losses = {
    Action.Scissors: [Action.Rock],  # Rock beats scissors
    Action.Rock: [Action.Paper],  # Paper beats rock
    Action.Paper: [Action.Scissors],  # Scissors beats paper
}


opponent_col1_map = {"A": Action.Rock, "B": Action.Paper, "C": Action.Scissors}
user_col2_map = {"X": Action.Rock, "Y": Action.Paper, "Z": Action.Scissors}

outcome_col2_map = {"X": "loss", "Y": "tie", "Z": "win"}


def determine_play(opponent_action: Action, outcome: str) -> Action:
    match outcome:
        case "tie":
            return_value = opponent_action
        case "win":
            return_value = losses[opponent_action][0]
        case "loss":
            return_value = victories[opponent_action][0]
        case _:
            raise "unknown play type"
    print(f"🧝‍♀️ wants a {outcome}, playing {return_value.name}")
    return return_value


def determine_score(user_action: Action, opponent_action: Action) -> int:
    # outcome of the round (0 if you lost, 3 if the round was a draw, and 6 if you won).

    defeats = victories[user_action]
    if user_action == opponent_action:
        print(f"Both 🧝 & 🧝‍♀️ selected {user_action.name}. It's a tie!")
        return 3 + user_action
    elif opponent_action in defeats:
        print(f"{user_action.name} beats {opponent_action.name}! 🧝‍♀️ wins, 🧝  loses!")
        return 6 + user_action
    else:
        print(f"{opponent_action.name} beats {user_action.name}! 🧝 wins,  🧝‍♀️ loses.")
        return user_action



score = 0
with open("../input.txt") as rps_data:
    for line in rps_data:
        opponent_symbol, outcome_symbol = line.strip().split()
        score += determine_score(
            determine_play(
                opponent_col1_map[opponent_symbol], outcome_col2_map[outcome_symbol]
            ),
            opponent_col1_map[opponent_symbol],
        )

print(f"\nTotal Score: {score}")

File: 02/go/day2-2.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

type Action uint64

const (
	Rock     Action = 1
	Paper    Action = 2
	Scissors Action = 3
)

func (a Action) String() string {
	switch a {
	case Rock:
		return "Rock"
	case Paper:
		return "Paper"
	case Scissors:
		return "Scissors"
	}
	return "Rock/Paper/Scissors"
}

var opponent_col1_map = map[string]Action{
	"A": Rock, "B": Paper, "C": Scissors}

var user_col2_map = map[string]Action{
	"X": Rock, "Y": Paper, "Z": Scissors}

var outcome_col2_map = map[string]string{
	"X": "loss", "Y": "draw", "Z": "win"}

var victories = map[Action]Action{
	Rock:     Scissors, // Rock beats scissors
	Paper:    Rock,     // Paper beats rock
	Scissors: Paper,    //	Scissors beats paper
}

var losses = map[Action]Action{
	Scissors: Rock,     // Rock beats scissors
	Rock:     Paper,    // Paper beats rock
	Paper:    Scissors, // Scissors beats paper
}

func determine_score(user_action Action, opponent_action Action) int {
	if user_action == opponent_action {
		fmt.Printf("Both 🧝 & 🧝‍♀️ selected %s. It's a tie!\n", user_action)
		return 3 + int(user_action)
	} else if opponent_action == victories[user_action] {
		fmt.Printf("%s beats %s! 🧝‍♀️ wins, 🧝 loses!\n", user_action, opponent_action)
		return 6 + int(user_action)
	} else {
		fmt.Printf("%s beats %s! 🧝 wins, 🧝‍♀️ loses.\n", opponent_action, user_action)
		return int(user_action)
	}
}

func determine_play(opponent_action Action, outcome string) Action {
	var return_value Action
	switch outcome {
	case "win":
		return_value = losses[opponent_action]
	case "loss":
		return_value = victories[opponent_action]
	default:
		return_value = opponent_action
	}
	fmt.Printf("🧝‍♀️ wants a %s, playing %s\n", outcome, return_value)
	return return_value
}

func main() {

	var score int = 0

	file, err := os.Open("../input.txt")

	if err != nil {
		log.Fatalf("failed to open")

	}
	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)

	for scanner.Scan() {
		each_ln := strings.TrimSpace(scanner.Text())
		symbols := strings.Split(each_ln, " ")

		score = score + determine_score(
			determine_play(opponent_col1_map[symbols[0]], outcome_col2_map[symbols[1]]),
			opponent_col1_map[symbols[0]])

	}
	file.Close()
	fmt.Println("Score:", score)
}
